src/commands/force/source/open.ts

```diff
@@ -0,0 +1,160 @@
+import checkLightningDomain from '@salesforce/core/lib/util/checkLightningDomain';
+import { SourceComponent, MetadataResolver } from '@salesforce/source-deploy-retrieve';
+import { OpenResultFormatter, OpenCommandResult } from '../../../formatters/openResultFormatter';
+
+export interface DnsLookupObject {
+  address: string;
+  family: number;
+}
+
+export interface FlexiPageRecord {
+  attributes: {
+    type: string;
+    url: string;
+  };
+  Id: string;
+}
+
+Messages.importMessagesDirectory(__dirname);
+const messages = Messages.loadMessages('@salesforce/plugin-source', 'open');
+
+export class Open extends SfdxCommand {
+  public static readonly description = messages.getMessage('description');
+  public static readonly examples = messages.getMessage('examples').split(os.EOL);
+  public static readonly requiresProject = true;
+  public static readonly requiresUsername = true;
+  public static readonly flagsConfig: FlagsConfig = {
+    sourcefile: flags.filepath({
+      char: 'f',
+      required: true,
+      description: messages.getMessage('SourceOpenFileDescription'),
+    }),
+    urlonly: flags.boolean({
+      char: 'r',
+      description: messages.getMessage('SourceOpenPathDescription'),
+    }),
+  };
+
+  protected openResult: OpenCommandResult;
+
+  public async run(): Promise<OpenCommandResult> {
+    await this.doOpen();
+    this.resolveSuccess();
+    return this.formatResult();
+  }
+
+  private async doOpen(): Promise<void> {
+    const typeName = this.getTypeNameDefinitionByFileName(path.resolve(this.flags.sourcefile));
+    const openPath = typeName === 'FlexiPage' ? await this.handleSupportedTypes() : await this.handleUnsupportedTypes();
+
+    this.openResult = await this.open(openPath);
+  }
+
+  private getTypeNameDefinitionByFileName(fsPath: string): string | undefined {
+    if (fs.fileExistsSync(fsPath)) {
+      const metadataResolver = new MetadataResolver();
+      const components: SourceComponent[] = metadataResolver.getComponentsFromPath(fsPath);
+      return components[0].type.name;
+    }
+    return undefined;
+  }
+
+  private async handleSupportedTypes(): Promise<string> {
+    return await this.setUpOpenPath();
+  }
+
+  private async handleUnsupportedTypes(): Promise<string> {
+    return await this.buildFrontdoorUrl();
+  }
+
+  private async getUrl(retURL: string): Promise<string> {
+    const frontDoorUrl: string = await this.buildFrontdoorUrl();
+    return `${frontDoorUrl}&retURL=${encodeURIComponent(decodeURIComponent(retURL))}`;
+  }
+
+  private async buildFrontdoorUrl(): Promise<string> {
+    const connection = this.org.getConnection();
+    const { username } = connection.getAuthInfoFields();
+    const authInfo = await AuthInfo.create({ username });
+    const url = authInfo.getOrgFrontDoorUrl();
+    return url;
+  }
+
+  private resolveSuccess(): void {
+    if (!getString(this.openResult, 'url')) {
+      process.exitCode = 1;
+    }
+  }
+
+  private isJsonOutput(): boolean {
```

Comment:
a few of these methods will be built into `SourceCommand`

Reply:
Done